std/numeric.d

```diff
@@ -3329,6 +3329,118 @@ C swapRealImag(C)(C input)
     return C(input.im, input.re);
 }
 
+/**
+Finds the mean (colloquially known as the average) of a range. If `r` does
+not provide the `length` member, then this function will do element by
+element summation, rather than the more accurate methods provided by
+$(REF sum, std,algorithm,iteration).
+
+An optional parameter `seed` is used for non-builtin types in order to
+correctly seed the summation with the equivalent to `0`.
+
+This function will return `real.nan` if the range is empty and if the element
+type of `r` is a built-in numerical type. However, if the seed value is defined,
+then empty ranges will return `ElementType!R.init`.
+
+This function is $(BIGOH r.length).
+
+Params:
+    r = An $(REF_ALTTEXT input range, isInputRange, std,range,primitives)
+    seed = An optional parameter for taking the mean of types which can't
+    implicitly convert to `real`. This value will initialize the summation,
+    so it should be equivalent to `0`.
+
+Returns:
+    the mean of r
+*/
+real mean(R)(R r)
+    if (isInputRange!R &&
+        is(ElementType!R : real) &&
+        !isInfinite!R)
+{
+    if (r.empty)
+    {
+        return real.nan;
+    }
+
+    return mean(r, real(0.0));
+}
+
+/// ditto
+auto mean(R, T)(R r, T seed)
+    if (isInputRange!R &&
+        (is(T == real) || !isNumeric!(T)) &&
+        is(typeof(seed + r.front)) &&
+        is(typeof(r.front / size_t(1))) &&
+        !isInfinite!R)
+{
+    import std.algorithm.iteration : sum, reduce;
+
+    if (r.empty)
+    {
+        return Unqual!(ElementType!R).init;
+    }
+
+    static if (hasLength!R)
+    {
+        if (r.length == 1)
+        {
+            return r.front;
+        }
+
+        return sum(r, seed) / r.length;
```

Comment:
> @JackStouffer Since std.las never materialized

FYI there's `mir.math.sum`: https://github.com/libmir/mir-algorithm/blob/master/source/mir/math/sum.d
It's well tested and highly optimized.